File: backend/app/services/scoring.py
from __future__ import annotations

from ..data import COMMUNITIES
from ..models import AnalyzeRequest, RankedCommunity
from .anchor import ResolvedAnchor, resolve_anchor
from .geo import clamp, haversine_miles
from .preferences import parse_preference_dimensions


def score(request: AnalyzeRequest) -> tuple[ResolvedAnchor, list[RankedCommunity]]:
    anchor = resolve_anchor(
        request.anchor_input,
        anchor_latitude=request.anchor_latitude,
        anchor_longitude=request.anchor_longitude,
        anchor_label=request.anchor_label,
    )
    dimensions = parse_preference_dimensions(request.lifestyle_preferences)

    household_weight = {
        "family": "family",
        "with pets": "pets",
        "couple": "quiet",
        "single": "nightlife",
    }[request.household]

    salary_budget = (request.salary / 12) * 0.34
    effective_budget = max(request.budget, salary_budget)

    ranked: list[RankedCommunity] = []

    for community in COMMUNITIES:
        if anchor.region != "custom" and community.region != anchor.region:
            continue

        distance_miles = haversine_miles(anchor.latitude, anchor.longitude, community.latitude, community.longitude)
        if distance_miles > request.radius:
            continue

        estimated_commute = distance_miles * 3.4 + 5
        if estimated_commute > request.commute_limit:
            continue
        commute_gap = abs(estimated_commute - request.commute_limit)
        commute_score = clamp(100 - commute_gap * 3.2, 0, 100)

        affordability_delta = community.avg_rent - effective_budget
        if affordability_delta > 0:
            continue
        affordability_score = clamp(100 - max(affordability_delta, 0) / 14, 12, 100)

        lifestyle_base = sum(getattr(community.lifestyle, dim) for dim in dimensions) / len(dimensions)
        weighted_lifestyle = clamp(
            lifestyle_base * 0.84 + getattr(community.lifestyle, household_weight) * 0.16,
            0,
            100,
        )

        overall_score = commute_score * 0.4 + affordability_score * 0.3 + weighted_lifestyle * 0.3

        ranked.append(
            RankedCommunity(
                id=community.id,
                name=community.name,
                region=community.region,
                latitude=community.latitude,
                longitude=community.longitude,
                distance_miles=round(distance_miles, 2),
                commute_score=round(commute_score, 2),
                affordability_score=round(affordability_score, 2),
                lifestyle_score=round(weighted_lifestyle, 2),
                overall_score=round(overall_score, 2),
                avg_rent=community.avg_rent,
            )
        )

    ranked.sort(key=lambda item: item.overall_score, reverse=True)
    return anchor, ranked[:10]

File: frontend/src/utils/scoring.ts
import type { HouseholdType, RankedCommunityApi } from '../api/analyze'
import { COMMUNITIES, type Community } from '../data/communities'
import type { RankedCommunity, ResolvedAnchor } from '../types/app'
import { clamp } from './format'
import { haversineMiles } from './geo'

export type PreferenceDimension = keyof Community['lifestyle']

export const parsePreferenceDimensions = (input: string): PreferenceDimension[] => {
  const normalized = input.toLowerCase()
  const matches: PreferenceDimension[] = []

  const keywordMap: Array<{ key: PreferenceDimension; words: string[] }> = [
    { key: 'walkable', words: ['walkable', 'walk', 'transit', 'dense'] },
    { key: 'quiet', words: ['quiet', 'calm', 'peaceful'] },
    { key: 'food', words: ['food', 'restaurant', 'foodie', 'coffee'] },
    { key: 'nightlife', words: ['nightlife', 'bar', 'club', 'party'] },
    { key: 'outdoors', words: ['outdoors', 'hiking', 'beach', 'park', 'nature'] },
    { key: 'family', words: ['family', 'kids', 'schools', 'child'] },
    { key: 'pets', words: ['pet', 'pets', 'dog', 'cat'] },
    { key: 'academic', words: ['academic', 'study', 'campus', 'student', 'university'] },
    { key: 'wellness', words: ['wellness', 'fitness', 'gym', 'health'] },
  ]

  keywordMap.forEach(({ key, words }) => {
    if (words.some((word) => normalized.includes(word))) {
      matches.push(key)
    }
  })

  return matches.length > 0 ? matches : ['walkable', 'food', 'quiet']
}

export const scoreCommunitiesLocally = ({
  anchor,
  budget,
  salary,
  commuteLimit,
  radius,
  lifestyleInput,
  household,
  communities = COMMUNITIES,
}: {
  anchor: ResolvedAnchor
  budget: number
  salary: number
  commuteLimit: number
  radius: number
  lifestyleInput: string
  household: HouseholdType
  communities?: Community[]
}): RankedCommunity[] => {
  const dims = parsePreferenceDimensions(lifestyleInput)
  const householdWeight =
    household === 'family' ? 'family' : household === 'with pets' ? 'pets' : household === 'couple' ? 'quiet' : 'nightlife'

  const monthlyAffordableFromSalary = (salary / 12) * 0.34
  const effectiveBudget = Math.max(budget, monthlyAffordableFromSalary)

  const candidates = communities
    .filter((community) => community.region === anchor.region || anchor.region === 'custom')
    .map((community) => {
      const distanceMiles = haversineMiles(anchor.latitude, anchor.longitude, community.latitude, community.longitude)
      if (distanceMiles > radius) {
        return null
      }

      const estimatedCommute = distanceMiles * 3.4 + 5
      if (estimatedCommute > commuteLimit) {
        return null
      }
      const commuteGap = Math.abs(estimatedCommute - commuteLimit)
      const commuteScore = clamp(100 - commuteGap * 3.2, 0, 100)

      const affordabilityDelta = community.avgRent - effectiveBudget
      if (affordabilityDelta > 0) {
        return null
      }
      const affordabilityScore = clamp(100 - Math.max(affordabilityDelta, 0) / 14, 12, 100)

      const lifestyleBase = dims.reduce((sum, key) => sum + community.lifestyle[key], 0) / dims.length
      const weightedLifestyle = clamp(lifestyleBase * 0.84 + community.lifestyle[householdWeight] * 0.16, 0, 100)

      const overallScore = commuteScore * 0.4 + affordabilityScore * 0.3 + weightedLifestyle * 0.3

      return {
        id: community.id,
        name: community.name,
        region: community.region,
        latitude: community.latitude,
        longitude: community.longitude,
        avgRent: community.avgRent,
        distanceMiles,
        commuteScore,
        affordabilityScore,
        lifestyleScore: weightedLifestyle,
        overallScore,
      }
    })
    .filter((community): community is RankedCommunity => community !== null)
    .sort((a, b) => b.overallScore - a.overallScore)

  return candidates.slice(0, 8)
}

export const toCommunity = (item: RankedCommunityApi): RankedCommunity => ({
  id: item.id,
  name: item.name,
  region: item.region,
  latitude: item.latitude,
  longitude: item.longitude,
  avgRent: item.avg_rent,
  distanceMiles: item.distance_miles,
  commuteScore: item.commute_score,
  affordabilityScore: item.affordability_score,
  lifestyleScore: item.lifestyle_score,
  overallScore: item.overall_score,
})

export const buildReason = (item: RankedCommunity): string => {
  if (item.lifestyleScore > 82) {
    return 'Strong lifestyle match for your stated preferences'
  }

  if (item.affordabilityScore > 82) {
    return 'Good cost fit relative to budget and salary'
  }

  if (item.commuteScore > 82) {
    return 'Commute window aligns closely with your target time'
  }

  return 'Balanced fit across commute, cost, and lifestyle factors'
}

export const buildTradeoff = (item: RankedCommunity): string => {
  const weakest = Math.min(item.commuteScore, item.affordabilityScore, item.lifestyleScore)
  if (weakest === item.affordabilityScore) {
    return 'Tradeoff: rent pressure is higher than your ideal level'
  }

  if (weakest === item.commuteScore) {
    return 'Tradeoff: commute may run longer during peak traffic'
  }

  return 'Tradeoff: fewer high-density amenities than core districts'
}
